src/b-splines/indexing.jl

```diff
@@ -158,9 +160,11 @@ end
 
 @generated function hessian(itp::AbstractInterpolation{T,N}, xs...) where {T,N}
     n = count_interp_dims(itp,N)
-    TH = promote_type(T, [x <: AbstractArray ? eltype(x) : x for x in xs]...)
     xargs = [:(xs[$d]) for d in 1:length(xs)]
-    :(hessian!(Array{TH, 2}($n,$n), itp, $(xargs...)))
+    quote
+        TH = $(Expr(:call, :promote_type, T, [x <: AbstractArray ? eltype(x) : x for x in xs]...))
```

Comment:
The problem is that this constructs an array and does the type computations at runtime.  It may not be type-stable, either.  It seems like you'd be better off doing something like `TH = hessian_type(T, xs...)` with
```jl
hessian_type(T) = T
hessian_type(T, x::AbstractArray, xs...) = promote_type(eltype(x), hessian_type(T, xs...))
hessian_type(T, x, xs...) = promote_type(typeof(x), hessian_type(T, xs...))
```

Reply:
Are you sure? That array is constructed inside the generator, not in the run-time code, and all the branching likewise happens inside the generator. I'm a fan of the lisp-y recursion trick anyway, but I'm pretty sure this case will work as-is.

Reply:
Oh, right, I misread your code — you're right, ther eshould continue to be no runtime cost here.